### PR TITLE
Finish writing+compiling the backprop of forward and backward.

### PR DESCRIPTION
**Pull request by danpovey**

### CODE REVIEW

**Comment by danpovey** [comment]
Merging for ease of testing etc.  Won't break anything.

File: k2/csrc/fsa_utils.cu [hunk 1]
```diff
@@ -1475,30 +1476,189 @@ Array1<FloatType> BackpropGetBackwardScores(
...
                     src_state_idx01 = fsas_row_ids2_data[fsas_arc_idx012],
                     state_idx0x = src_state_idx01 - src_state_idx1,
                     dest_state_idx01 = state_idx0x + dest_state_idx1;
+            FloatType src_score = forward_scores_data[src_state_idx01],
+                      arc_end_score = src_score + arc.score,
+                      dest_score = forward_scores_data[dest_state_idx01];
+            // alpha = d(dest_score) / d(arc_end_score)
+            FloatType alpha = exp(arc_end_score - dest_score),
+                      arc_deriv =
+                          alpha * forward_scores_deriv_data[src_state_idx01];
```
Comment (csukuangfj) [inline k2/csrc/fsa_utils.cu:1591]: Should it be
```cpp
alpha * forward_scores_deriv_data[dest_state_idx01];
```
?
Correct me if I am wrong.

Comment (danpovey) [reply k2/csrc/fsa_utils.cu:1591]: yes you are right.  @qindazhu can you fix in next commit?

Comment (qindazhu) [reply k2/csrc/fsa_utils.cu:1591]: I have fixed it locally

**Pull request merged**